_posts/2020-10-12-sparklyr-flint-0.2.0-released/sparklyr-flint-0.2.0.Rmd

```diff
@@ -0,0 +1,143 @@
+##  6 1970-01-01 00:00:06     6     7
+##  7 1970-01-01 00:00:07     7     8
+##  8 1970-01-01 00:00:08     8     9
+##  9 1970-01-01 00:00:09     9    10
+## 10 1970-01-01 00:00:10    10    11
+```
+
+Notice regardless of which temporal direction is selected, an outer-left join is always performed (i.e., all timestamp values and `u` values of `left` from above will always be present in the output, and the `v` column in the output will contain `NA` whenever there is no record from `right` that meets the matching criteria).
+
+## OLS Regression
+
+You might be wondering whether the version of this functionality in Flint is more or less identical to `lm()` in R. Turns out it has much more to offer than `lm()` does. An OLS regression in Flint will compute useful metrics such as [Akaike information criterion](https://en.wikipedia.org/wiki/Akaike_information_criterion) and [Bayesian information criterion](https://en.wikipedia.org/wiki/Bayesian_information_criterion), both of which are useful for model selection purposes, and the calculations of both are parallelized by Flint to fully utilize computational power available in a Spark cluster. In addition, Flint supports ignoring regressors that are constant or nearly constant, which becomes useful when an intercept term is included. To see why this is the case, we need to briefly examine the goal of the OLS regression, which is to find some column vector of coefficients $\mathbf{\beta}$ that minimizes $\|\mathbf{y} - \mathbf{X} \mathbf{\beta}\|^2$, where $\mathbf{y}$ is the column vector of response variables, and $\mathbf{X}$ is a matrix consisting of columns of regressors plus an entire column of $1$s representing the intercept terms. The solution to this problem is $\mathbf{\beta} = (\mathbf{X}^\intercal\mathbf{X})^{-1}\mathbf{X}^\intercal\mathbf{y}$, assuming the Gram matrix $\mathbf{X}^\intercal\mathbf{X}$ is non-singular. However, if $\mathbf{X}$ contains a column of all $1$s of intercept terms, and another column formed by a regressor that is constant (or nearly so), then columns of $\mathbf{X}$ will be linearly dependent (or nearly so) and $\mathbf{X}^\intercal\mathbf{X}$ will be singular (or nearly so), which presents an issue computation-wise. However, if a regressor is constant, then it essentially plays the same role as the intercept terms do. So simply excluding such a constant regressor in $\mathbf{X}$ solves the problem. Also, speaking of inverting the Gram matrix, readers remembering the concept of "condition number" from numerical analysis must be thinking to themselves how computing $\mathbf{\beta} = (\mathbf{X}^\intercal\mathbf{X})^{-1}\mathbf{X}^\intercal\mathbf{y}$ could be numerically unstable if $\mathbf{X}^\intercal\mathbf{X}$ has a large condition number. This is why Flint also outputs the condition number of the Gram matrix in the OLS regression result, so that one can sanity-check the underlying quadratic minimization problem being solved is well-conditioned.
```

Comment:
Personally I'd remove the allusion to `lm`, lest someone thinks the  comparison is unfair :-)

(I mean - `lm` is like R's "hello world" ... at least that's what I called in a talk once ;-)))

If you leave it in, did you try what actually happens if you have a constant predictor in `lm`? (Probably will also just silently remove it?)

Reply:
`lm` outputs NA as coefficients if the Gram matrix is singular.

Minimal example:

```
df <- data.frame(x = c(1, 1), y = c(1, 2))
lm(formula = y ~ x + 1, data = df)

# outputs NA as coefficient for x
```

vs

```
df <- data.frame(x = c(1, 2), y = c(1, 2))
lm(formula = y ~ x + 1, data = df)

# Gram matrix is not singular so output looks normal
```